server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -598,5 +412,97 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
+         * The {@link ForIndexDimensions} strategy avoids this double hashing.
+         */
+        public static class ForRoutingPath extends ExtractFromSource {
+            private final Predicate<String> isRoutingPath;
+
+            ForRoutingPath(IndexMetadata metadata, List<String> routingPath) {
+                super(metadata, routingPath);
+                isRoutingPath = Regex.simpleMatcher(routingPath.toArray(String[]::new));
+            }
+
+            @Override
+            protected int hashSource(IndexRequest indexRequest) {
+                return hashRoutingFields(indexRequest.getContentType(), indexRequest.source()).buildHash(
+                    IndexRouting.ExtractFromSource::defaultOnEmpty
+                );
+            }
+
+            public String createId(XContentType sourceType, BytesReference source, byte[] suffix) {
+                return hashRoutingFields(sourceType, source).createId(suffix, IndexRouting.ExtractFromSource::defaultOnEmpty);
+            }
+
+            public RoutingHashBuilder builder() {
+                return new RoutingHashBuilder(isRoutingPath);
+            }
+
+            private RoutingHashBuilder hashRoutingFields(XContentType sourceType, BytesReference source) {
+                RoutingHashBuilder b = builder();
+                try (XContentParser parser = XContentHelper.createParserNotCompressed(parserConfig, source, sourceType)) {
+                    parser.nextToken(); // Move to first token
+                    if (parser.currentToken() == null) {
+                        throw new IllegalArgumentException("Error extracting routing: source didn't contain any routing fields");
+                    }
+                    parser.nextToken();
+                    b.extractObject(null, parser);
+                    ensureExpectedToken(null, parser.nextToken(), parser);
+                } catch (IOException | ParsingException e) {
+                    throw new IllegalArgumentException("Error extracting routing: " + e.getMessage(), e);
+                }
+                return b;
+            }
+
+            public boolean matchesField(String fieldName) {
+                return isRoutingPath.test(fieldName);
+            }
+        }
+
+        /**
+         * Strategy for time-series indices that use {@link IndexMetadata#INDEX_DIMENSIONS} to extract the tsid from the source.
+         * This strategy avoids double hashing of dimensions during indexing.
+         * It requires that the index was created with {@link IndexVersions#TSID_CREATED_DURING_ROUTING} or later.
+         * It creates the tsid during routing and makes the routing decision based on the tsid.
+         * The tsid gets attached to the index request so that the data node can reuse it instead of rebuilding it.
+         */
+        static class ForIndexDimensions extends ExtractFromSource {
+
+            ForIndexDimensions(IndexMetadata metadata, List<String> indexDimensions) {
```

Comment:
Nit: similarly here I think we could extract the dimensions inside the constructor based on metadata.